tensorflow/contrib/opt/python/training/model_average_device_setter.py

```diff
@@ -0,0 +1,147 @@
+        spec.merge_from(current_device)
+        return spec.to_string()
+      else:
+        return device_string
+    else:
+      if self._worker_device:
+        if not self._merge_devices:
+          return self._worker_device
+        spec = pydev.DeviceSpec.from_string(self._worker_device)
+
+      if not self._merge_devices:
+        return ""
+
+    spec.merge_from(current_device)
+    return spec.to_string()
+
+def model_average_device_setter(ps_tasks=0, ps_device="/job:ps",
+                          worker_device="/job:worker", merge_devices=True,
+                          cluster=None, ps_ops=None, ps_strategy=None):
+  """Return a `device function` to use when building a Graph for model average.
+
+  There is only one difference between model_average_device_setter and
+  replica_device_setter : replica_device_setter placed all variables
+  (including global/local variables) on ps, while in model average, each worker
+  own its local variables (local model parameters), these local variables
+  should be placed in each worker.
+  Args:
+    ps_tasks: Number of tasks in the `ps` job.  Ignored if `cluster` is
+      provided.
+    ps_device: String.  Device of the `ps` job.  If empty no `ps` job is used.
+      Defaults to `ps`.
+    worker_device: String.  Device of the `worker` job.  If empty no `worker`
+      job is used.
+    merge_devices: `Boolean`. If `True`, merges or only sets a device if the
+      device constraint is completely unset. merges device specification rather
+      than overriding them.
+    cluster: `ClusterDef` proto or `ClusterSpec`.
+    ps_ops: List of strings representing `Operation` types that need to be
+      placed on `ps` devices.  If `None`, defaults to `["Variable"]`.
+    ps_strategy: A callable invoked for every ps `Operation` (i.e. matched by
+      `ps_ops`), that takes the `Operation` and returns the ps task index to
+      use.  If `None`, defaults to a round-robin strategy across all `ps`
+      devices.
+
+  Returns:
+    A function to pass to `tf.device()`.
+
+  Raises:
+    TypeError if `cluster` is not a dictionary or `ClusterDef` protocol buffer,
+    or if `ps_strategy` is provided but not a callable.
+  """
+  if cluster is not None:
+    if isinstance(cluster, server_lib.ClusterSpec):
+      cluster_spec = cluster.as_dict()
+    else:
+      cluster_spec = server_lib.ClusterSpec(cluster).as_dict()
+    # Get ps_job_name from ps_device by striping "/job:".
+    ps_job_name = pydev.DeviceSpec.from_string(ps_device).job
+    if ps_job_name not in cluster_spec or cluster_spec[ps_job_name] is None:
+      return None
+    ps_tasks = len(cluster_spec[ps_job_name])
+
+  if ps_tasks == 0:
+    return None
+
+  if ps_ops is None:
+    ps_ops = ["Variable", "VariableV2", "VarHandleOp"]
+
+  if not merge_devices:
+    logging.warning(
+        "DEPRECATION: It is recommended to set merge_devices=true in "
+        "replica_device_setter")
```

Comment:
Please don't add deprecation warnings to new code. Should one be using `merge_devices` or not? If it should be true, remove the parameter.